tools/config.py

```diff
@@ -489,28 +502,90 @@ def regions(self):
                                       "start at 0x%x" % rom_start)
-            part_size = (part.maxaddr() - part.minaddr()) + 1
-            yield Region("bootloader", rom_start + start, part_size, False,
-                         filename)
-            start += part_size
+            bootloader_size = (bootloader_part.maxaddr() - bootloader_part.minaddr()) + 1
+
+            # yeild bootloader region
+            yield Region("bootloader", bootloader_start, bootloader_size, False, bootloader_fn)
+
+            # update rom pointer
+            rom_current_addr = bootloader_start + bootloader_size
+
+            # find metadata header region start
+            metadata_start = rom_current_addr
+            if metadata_start%FLASH_ERASE_SIZE: # align to erase boundary
+                metadata_start = ((metadata_start // FLASH_ERASE_SIZE) + 1) * FLASH_ERASE_SIZE
+
+            # find metadata header size
+            metadata_size = FIRMWARE_HEADER_SIZE
+
+            # yeild image_metadata_header region
+            yield Region("firmware_metadata_header", metadata_start, metadata_size, False, None)
+
+            # update ROM pointer
+            rom_current_addr = metadata_start + metadata_size
+
+        # Application Region
+        # find application start address
+        app_start = rom_current_addr
+        if app_start%VECTOR_TABLE_SIZE: # align to vector table size
+            app_start = ((app_start // VECTOR_TABLE_SIZE) + 1) * VECTOR_TABLE_SIZE
+
+        # find application size
+        app_size = rom_start + rom_size - app_start
         if 'target.restrict_size' in target_overrides:
-            new_size = int(target_overrides['target.restrict_size'], 0)
-            yield Region("application", rom_start + start, new_size, True, None)
-            start += new_size
-            yield Region("post_application", rom_start +start, rom_size - start,
-                         False, None)
-        else:
-            yield Region("application", rom_start + start, rom_size - start,
-                         True, None)
-        if start > rom_size:
-            raise ConfigException("Not enough memory on device to fit all "
-                                  "application regions")
+            restrict_size = int(target_overrides['target.restrict_size'], 0)
+            app_size = restrict_size if restrict_size < app_size else app_size
+
+        # yield application region
+        yield Region("application", app_start, app_size, True, None)
+
+        # update ROM pointer
+        rom_current_addr = app_start + app_size
+
+        # yeild the metadata region to expose symbols to help build the bootloader
+        if self.building_bootloader:
+            # find metadata header region start
+            metadata_start = rom_current_addr
+            if metadata_start%FLASH_ERASE_SIZE: # align to erase boundary
+                metadata_start = ((metadata_start // FLASH_ERASE_SIZE) + 1) * FLASH_ERASE_SIZE
```

Comment:
Can we re-factor this "round up to nearest block" function out? it might be generally useful.